src/dawn/SIR/ASTNodeIterator.h

```diff
@@ -0,0 +1,308 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_SIR_ASTASTNODEITERATOR_H
+#define DAWN_SIR_ASTASTNODEITERATOR_H
+
+#include "ASTStmt.h"
+#include "dawn/Support/IteratorRange.h"
+#include <list>
+#include <type_traits>
+
+namespace dawn {
+
+/// @brief (Abstract) Pre-order visiting iterator through the Stmts of one or more ASTs.
+/// Specializations should specify on which ASTs to iterate and in which order.
+/// @ingroup sir
+template <typename GeneratorType, bool onlyFirstLevel, typename Enable = void>
```

Comment:
After the discussion on comments below, I am not sure we need this base class. But if so, I would replace Enable
by a static_assert inside the body. I would use the enable if, to dispatch two different specializations

Reply:
I would keep the base class in case one wants to extend it for different `GeneratorType`s (such as IIRNodes as we were considering before).
I am not sure I understand the other part of your comment, could you provide an example?

Now it occurs to me that having specializations such as `ASTNodeIterator<BlockStmt, true/false>` or `ASTNodeIterator<IfStmt, true/false>` ..., despite being more expressive and explicit, could be counterproductive, as you have no way to cast it to an `ASTNodeIterator<Stmt, true/false>` which is polymorphically valid for all kinds of statements and, also, these specializations don't have actual use cases.
So we could consider having just one specialization of `ASTNodeIterator<...>`, which is `ASTNodeIterator<Stmt, onlyFirstLevel>`. This way, we would also avoid the whole `enable_if` - `is_base_of` construct which certainly makes the code unreadable (and causes quite some headaches).

Reply:
Last commit removes everything mentioned.